File: MainPart/src/day03/303.py
from typing import List

class NumArray:

    def __init__(self, nums: List[int]):
        self.nums = nums
        self.sum_array = [0] + [sum(nums[0:i+1:1]) for i in range(len(nums))]
        

    def sumRange(self, left: int, right: int) -> int:
        return self.sum_array[right] - self.sum_array[left]        


# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# param_1 = obj.sumRange(left,right)

File: MainPart/src/day03/303.c
#include <stdlib.h>


typedef struct {
    int* content;
    int* sum_array;
    int numsize;
} NumArray;


NumArray* numArrayCreate(int* nums, int numsSize) {
    // NumArray* num_array;
    // 这是不对的，因为这样是野指针
    NumArray* num_array = (NumArray *) calloc(1, sizeof(NumArray));

    num_array -> content = (int *) calloc(numsSize, sizeof(int));
    num_array -> sum_array = (int *) calloc(numsSize+1, sizeof(int));
    num_array -> numsize = numsSize;
    int tmp = 0;

    for (int i = 0; i < numsSize; i++) {
        num_array -> content [i] = nums [i];
        tmp += nums[i];
        num_array -> sum_array[i+1] = tmp;
    }
    return num_array;
}

int numArraySumRange(NumArray* obj, int left, int right) {
    return (obj->sum_array[right+1] - obj->sum_array[left]);
}

void numArrayFree(NumArray* obj) {
    free(obj -> content);
    free(obj -> sum_array);
    free(obj);
    return ;
}

/**
 * Your NumArray struct will be instantiated and called as such:
 * NumArray* obj = numArrayCreate(nums, numsSize);
 * int param_1 = numArraySumRange(obj, left, right);
 
 * numArrayFree(obj);
*/